planner/cascades/implementation.go

```diff
@@ -0,0 +1,48 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cascades
+
+import (
+	plannercore "github.com/pingcap/tidb/planner/core"
+)
+
+// Implementation defines the interface for cost of physical plan.
+type Implementation interface {
+	calcCost(outCount float64, childCosts []float64, children ...*Group) float64
+	setCost(cost float64)
+	getCost() float64
+	getPlan() plannercore.PhysicalPlan
+}
+
+type baseImplementation struct {
+	cost float64
+	plan plannercore.PhysicalPlan
+}
+
+func (impl *baseImplementation) calcCost(outCount float64, childCosts []float64, children ...*Group) float64 {
```

Comment:
How about storing self cost and the cumulative cost in `baseImplementation`, changing this function to:
```go
func (impl *baseImplementation) calcSelfCost(logicalProp *property.LogicalProperty, children ...*Group) float64 {
	// calculate self cost, store it in impl.selfCost
	// return impl.selfCost
}
```

The cumCost of `impl` can be initialized with:
```go
impl.cumCost = impl.cumCost
```

and in https://github.com/pingcap/tidb/pull/8449/files#diff-46462632826dca717f2913bf43b93625R101, maintain the cumulative cost by:
```gp
impl.cumCost += childImpl.cumCost
```

and the `cumCost` in https://github.com/pingcap/tidb/pull/8449/files#diff-46462632826dca717f2913bf43b93625R88 can be initialized with `impl.selfCost`